### PR TITLE
fix(codecs): Improve deserialization of `EncodingConfigAdapter` / fix overriding `framing`

### LINKED ISSUE #12473: Poor error message when missing required encoding configuration for `aws_kinesis_firehose`
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Not providing encoding configuration for the `aws_kinesis_firehose` sink now returns:

```
2022-04-28T22:55:54.593972Z ERROR vector::cli: Configuration error. error=data did not match any variant of untagged enum EncodingConfigAdapter for key `sinks.sink0`
```

I believe as of https://github.com/vectordotdev/vector/pull/12176

We should add some tests around this when fixing it. I'm guessing other sinks that have been converted might be affected by this too.

### Configuration

_No response_

### Version

vector 0.22.0 (aarch64-apple-darwin debug=full)

### Debug Output

_No response_

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by pablosichert**
Closes #12473.

The error messages are improved in so far that sinks will reject unknown fields again and missing `framing` and `encoding` fields are mentioned explicitly.

Unfortunately, deserializing the `encoding` field itself will still give vague error messages due to https://github.com/serde-rs/serde/pull/1544 (some more details in #12162).

Additionally, this PR fixes a bug where the `framing` config could not be overridden when a legacy codec was used.

### CODE REVIEW

**Review by JeanMertz — APPROVED** [review]
That's some interesting adapter-fu going on there 👀 😄 

It all _looks_ okay to me, but the diff is hard to follow, so I'd wait for one more pair of eyes.

**Comment by pablosichert** [comment]
> That's some interesting adapter-fu going on there 👀 😄

Yeah, it's... involved. Scimming the tests is probably a good start:

https://github.com/vectordotdev/vector/blob/af7514b8e8e64b893138fdf249ac235cb82ac9b6/src/sinks/util/encoding/adapter.rs#L500-L661

**Review by jszwedko — APPROVED** [review]
Thanks for the docs note!
